File: Data Structures & Algorithms/max-water-container/submission-3.java
class Solution {
    public int maxArea(int[] heights) {
        int l = 0, r = heights.length - 1;
        int max = 0;
        int curr;
        while (l < r) {
            curr = (r - l) * Math.min(heights[l], heights[r]);
            if(curr > max) max = curr;
            if(heights[l] < heights[r]) {
                l++;
                continue;
            } else {
                r--;
                continue;
            }
        }
        return max;
    }
}

File: Data Structures & Algorithms/maximum-product-subarray/submission-3.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        minVal = [float("inf")] * len(nums)
        maxVal = [-float("inf")] * len(nums)

        minVal[0] = maxVal[0] = nums[0]
        for i in range(1, len(nums)):
            minVal[i] = min(nums[i], minVal[i - 1] * nums[i], maxVal[i - 1] * nums[i])
            maxVal[i] = max(nums[i], minVal[i - 1] * nums[i], maxVal[i - 1] * nums[i])
        return max(maxVal)